plugin/types.go

```diff
@@ -80,11 +83,27 @@ type Parent interface {
 	List(context.Context) ([]Entry, error)
 }
 
+// SchemaMap represents a map of <type> => <JSON schema>.
+type SchemaMap = map[interface{}]*JSONSchema
+
 // Root represents the plugin root. The Init function is passed a config map representing
-// plugin-specific configuration.
+// plugin-specific configuration. WrappedTypes returns a map of <type> => <JSONSchema>.
+// It is used by the EntrySchema#SetMeta*Schema methods to return the right metadata schema
+// for wrapped types. You should only implement WrappedTypes if your plugin's API wraps primitive
+// types like date, integer, number, string, boolean, etc. Otherwise, WrappedTypes should return
+// nil. See kubernetes/root.go for an example of when WrappedTypes should not return nil.
+//
+// NOTE: Type aliases like "type Time = time.Time" do NOT count as wrapped types. Pointers
+// also don't count. Only promoted types ("type Time time.Time") and wrapper structs
+// ("struct Time { t time.Time }") count.
+//
+// NOTE: Without WrappedTypes, the underlying JSON schema library will treat promoted types and
+// wrapper structs as JSON objects, which would be incorrect. This, unfortunately, is a
+// limitation of Go's reflect package.
 type Root interface {
 	Parent
 	Init(map[string]interface{}) error
+	WrappedTypes() SchemaMap
```

Comment:
Another alternative I was considering was making this optional. E.g. something like
```
type HasWrappedTypes struct {
    Root
    WrappedTypes() SchemaMap
}
```

That way, plugin authors whose SDKs don't have potentially conflicting wrapped types won't have to implement this. However, that is also an extra thing we'll need to document, which is why I like having `WrappedTypes()` in Root. Although returning nil can be redundant, it's easy enough (and done in only one place -- the root) that we may as well require plugin authors to implement it.

I'm also not sure if WrappedTypes is a good name, so open to other alternatives.

Reply:
We need to document it either way. I lean towards making it optional, we've already added lots of required functions that make it more complicated to start.

Reply:
Ok. Any suggestions on the name?

Reply:
What you proposed seems fine.